src/lakefs_spec/transaction.py

```diff
@@ -0,0 +1,144 @@
+from __future__ import annotations
+
+from collections import deque
+from dataclasses import dataclass
+from functools import partial
+from typing import TYPE_CHECKING, Any, Callable, Generic, TypeVar
+
+from fsspec.spec import AbstractBufferedFile
+from fsspec.transaction import Transaction
+from lakefs_sdk.client import LakeFSClient
+from lakefs_sdk.models import Commit
+
+from lakefs_spec.client_helpers import commit, create_tag, ensure_branch, merge, rev_parse, revert
+
+T = TypeVar("T")
+
+if TYPE_CHECKING:
+    from lakefs_spec import LakeFSFileSystem
+
+    VersioningOpTuple = tuple[Callable[[LakeFSFileSystem], None], Any]
+
+
+@dataclass
+class Placeholder(Generic[T]):
+    value: T | None = None
+
+    def available(self):
+        return self.value is not None
+
+    def set_value(self, value: T) -> None:
+        self.value = value
```

Comment:
Clarifying question: Where does the `set_value` function get called? Is it under the hood of `fs-spec`, or when does the replacement of the placeholders with the values happen?

Reply:
That happens here: 

https://github.com/appliedAI-Initiative/lakefs-spec/pull/140/files#diff-6244f5c90acecfad8e9a92a83e64c21a31bfab360bd819eb87b17f04fd63f7c8R100-R101

In words, placeholders are immediate return values of the transaction versioning helpers (i.e. `tx.commit, tx.rev_parse`), and are stored in the `files` queue together with the `partial`-wrapped client helper.

On unwinding the queue in the transaction completion, the return value of the client helper partial (i.e., the actual computed value which we were waiting for) is assigned to the placeholder.